### PR TITLE
Children

### PR DESCRIPTION
**Pull request by andreirtaylor**
This was almost as much of a nightmare as actually having children.

Adding children works as expected now as far as I can tell

This should be merged then we can deal with #87

image

### CODE REVIEW

**Comment by cpjk** [comment]
needs a rebase

File: views/edit_or_add_mixin.pug [hunk 1]
```diff
@@ -0,0 +1,46 @@
+mixin editScript
+  // this needs to be a pug mixin because it transfers the children in a really stupid way
+  script(type='text/javascript').
+    // this is the stupidity im talking about
+    var children = [];
+    var chorals = !{JSON.stringify(chorals)};
+
+    function abortChild(index){
```
Comment (cpjk) [inline views/edit_or_add_mixin.pug]: > abort child

dayum

**Comment by cpjk** [comment]
it would be sweet if on `/chorals/edit/:choral_id`, the current children of the choral are pre-populated on the page, so i dont have to remember what its children were

**Comment by cpjk** [comment]
other than that lgtm

**Comment by cpjk** [comment]
> it would be sweet if on /chorals/edit/:choral_id, the current children of the choral are pre-populated on the page, so i dont have to remember what its children were

I say merge it after rebase then add that in a separate PR

**Comment by andreirtaylor** [comment]
They were doing that before I changed it.

**Comment by andreirtaylor** [comment]
Sloppy on my part

**Comment by andreirtaylor** [comment]
Also I fixed the children issue. They are now shown on the edit page

File: routes/chorals.js [hunk 2]
```diff
@@ -136,7 +137,7 @@ router.get('/:choralId', function(req, res, next) {
     var ret = {};
     Choral.findOne({ choralId: choralId }, (err, choral) => {
       if (err || !choral) {
-        logHelper.createLog("error", 'Choral does not exist: ' + err, ["chorals", "get"]);
+        logHelper.createLog("error", 'Choral does not exist: ' + err, ["chorals", "delete"]);
```
Comment (cpjk) [inline routes/chorals.js]: why delete

File: routes/chorals.js [hunk 3]
```diff
@@ -147,13 +148,12 @@ router.get('/:choralId', function(req, res, next) {
...
     });
-  });
-
+  }); 
   p.then((ret) => {
     return new Promise((resolve,reject) => {
       client.hgetall(choralId, (err, data) => {
         if(err) {
-          logHelper.createLog("error", 'Choral data has not yet been published to redis: ' + err, ["chorals", "get"]);
+          logHelper.createLog("error", 'Choral data has not yet been published to redis: ' + err, ["chorals", "delete"]);
```
Comment (cpjk) [inline routes/chorals.js]: why delete

File: routes/chorals.js [hunk 4]
```diff
@@ -242,7 +242,7 @@ router.delete('/:choralId', function(req, res, next) {
       }
 
       res.flash('success', 'Choral successfully deleted.');
-      return res.send('204'); // successful deletion
+      return res.send('204'); // notify client of failure
```
Comment (cpjk) [inline routes/chorals.js]: not failure

File: routes/chorals.js [hunk 5]
```diff
@@ -327,16 +303,17 @@ router.post('/edit/:choralId', function(req, res, next) {
...
       res.flash('error', 'Choral does not exist');
       return res.send('404'); // notify client of failure
     }
-    choral.edit(attrs, (err, choral) => {
+    choral.edit(attrs, (err) => {
       if(err){
         console.log(err);
         res.flash('error', 'Choral validation failed: ' + err);
-        return next(err);
+        console.log(req.originalUrl )
```
Comment (cpjk) [inline routes/chorals.js]: maybe :fire: this logging

File: views/newChoral.pug [hunk 6]
```diff
@@ -7,7 +8,7 @@ block content
       b
         i.fa.fa-plus-square-o
         |  Create A New Choral
-  form.w3-container(method='POST' action='/chorals' id='mainForm').new-choral-form
+  form.w3-container(method='POST' action='/chorals/new' id='mainForm').new-choral-form
```
Comment (cpjk) [inline views/newChoral.pug]: why POSTing to /new?

File: routes/chorals.js [hunk 7]
```diff
@@ -38,18 +37,26 @@ router.get('/', function(req, res, next) {
...
+      reqChildren = [ reqChildren ];
+  }
+  var children = [];
+  if(reqChildren && reqChildren.length){
+      children = reqChildren.map((val) => {return { _id: val } } );
+  }
+  return children;
+}
+
+router.post('/new', function(req, res, next) {
```
Comment (cpjk) [inline routes/chorals.js]: why did you change this to POSTing to /new?

File: views/editChoral.pug [hunk 8]
```diff
@@ -7,33 +8,24 @@ block content
       b
         i.fa.fa-dashboard
         |  Edit Choral
-  form.w3-container(method='POST' action='/chorals/edit/'+choralEdit.choralId id='mainForm').new-choral-form
+  form.w3-container(method='POST' action='/chorals/edit/'+ choralEdit.choralId id='mainForm').new-choral-form
```
Comment (james-woo) [inline views/editChoral.pug:11]: Could you add another form group, if you think it is a good idea:
.form-group
  label(for='choralIdInput') ID
  input#choralIdInput.form-control(name='id' type='text' readonly value=choralEdit.choralId)

File: routes/chorals.js (hunk 7 above)
Comment (andreirtaylor) [reply routes/chorals.js]: To me it makes more sense to post to new. I don't really get why it was posting to nowhere in the first place

File: views/editChoral.pug (hunk 8 above)
Comment (andreirtaylor) [reply views/editChoral.pug:11]: I was also wondering about that. I think it would be a bit cleaner to do it your way but the infrastructure was already there do do it with a url parameter.

Comment (andreirtaylor) [reply views/editChoral.pug:11]: If you think it's a good idea let me knno

File: routes/chorals.js (hunk 7 above)
Comment (cpjk) [reply routes/chorals.js]: Just a convention. POST /resource usually means create that resource

Comment (andreirtaylor) [reply routes/chorals.js]: Yeah I can change it back. I had no idea that was a convention

Comment (cpjk) [reply routes/chorals.js]: 👌 yeah, it's just a convention of REST.

Comment (cpjk) [reply routes/chorals.js]: Of course, we don't necessarily need to follow it, but we already started 👷

File: views/editChoral.pug (hunk 8 above)
Comment (james-woo) [reply views/editChoral.pug:11]: I think its a good idea, so that the user can see the choralId

**Comment by cpjk** [comment]
Noticed one more thing. Adding children on choral creation does not appear to work.

File: views/editChoral.pug (hunk 8 above)
Comment (andreirtaylor) [reply views/editChoral.pug:11]: Done

**Comment by cpjk** [comment]
There are still conflicts with master @andreirtaylor

File: models/choral.js [hunk 9]
```diff
@@ -9,13 +9,32 @@ var choralSchema = new mongoose.Schema({
...
+          },
+          message: 'Device-type chorals are not allowed to have children!'
+        },
+        {
+          isAsync: true,
+          validator: function(v, cb) { // ensure devices have no children
+            if(!v) return true;
+
+            Choral.find({_id: {$in: v }}, (err, found) => {
+              console.log(found, v)
```
Comment (cpjk) [inline models/choral.js]: :fire: logging

File: routes/chorals.js [hunk 10]
```diff
@@ -251,71 +251,47 @@ router.get('/edit/:choralId', function(req, res, next) {
...
-  var cookies = req.get("Cookie");
-  var children = [];
-  var child;
-
-  for(var i = 0; cookieHelper.readCookie('child' + i, cookies) != null; i++){
-    child = cookieHelper.readCookie('child' + i, cookies);
-    child = JSON.parse(child);
-    children.push(child.choralId);
-  }
+  console.log(req.body)
```
Comment (cpjk) [inline routes/chorals.js]: :fire: logging

Comment (andreirtaylor) [reply routes/chorals.js]: Jeez how did I miss that

**Comment by cpjk** [comment]
thumbs_up
